service/consul/environment.go

```diff
@@ -209,13 +208,72 @@ func NewEnvironment(l log.Logger, registrationScheme string, co Options, eo ...s
 		return nil, err
 	}
 
-	return environment{
+	newServiceEnvironment := environment{
 		service.NewEnvironment(
 			append(
 				eo,
 				service.WithRegistrars(r),
 				service.WithInstancers(i),
 				service.WithCloser(closer),
-			)...,
-		), NewClient(consulClient)}, nil
+			)...), NewClient(consulClient)}
+
+	if co.DatacenterWatchInterval > 0 {
+		go WatchInstancers(l, co, newServiceEnvironment)
+	}
+
+	return newServiceEnvironment, nil
+}
+
+func WatchInstancers(l log.Logger, co Options, e Environment) {
+	if co.DatacenterWatchInterval <= 0 {
+		l.Log(level.Key(), level.WarnValue(), logging.MessageKey(), "Not setting up instancer watch.", "Watch interval: ", co.DatacenterWatchInterval)
+		return
+	}
+
+	createInstancersAgain := time.NewTicker(co.DatacenterWatchInterval)
+	client := e.Client()
+	for {
+		<-createInstancersAgain.C
+
+		currentInstancers := e.Instancers()
+		keys := make(map[string]bool)
+		instancersToAdd := make(service.Instancers)
+
+		var datacenters []string
+		var err error
+		for _, w := range co.watches() {
+			if w.CrossDatacenter {
+				if len(datacenters) == 0 {
+					datacenters, err = getDatacenters(l, client, co)
```

Comment:
I think we'd want to get a fresh copy every time in case a new dc was added.

Reply:
It was my understanding that `getDatacenters` gets the datacenters that are part of the consul client's catalog, which is updated whenever a datacenter is added or removed right?

Reply:
@j-mai  is correct `getDatacenters` returns all the current datacenters.

Reply:
So wouldn't we want to make that call at every tick of `createInstancesAgain`? Say updating line 242 to 
```datacenters, err := getDatacenters(l, client, co)```
I just didn't quite understand the logic behind line 246.

Reply:
Yeah, that makes sense. I think my reasoning behind adding line 246 is in case `getDatacenters` errors, then maybe we want to try again on the next watch that we are looping through. But I think it does make more sense to just get it at line 242 and if it errors, wait until `createInstancesAgain` ticks again to check for the datacenters again, since it's likely that getting the datacenters on the next watch will just error again.

Reply:
Exactly, it'll keep the code simple and readable 😃  And if we see the need to add retries (unless the call already performs retries under the hood) later on, we can definitely do that